File: swift/Sources/CircleAI/PersonalFinance.swift
// PersonalFinance.swift
//
// Port of the Personal.Finance vertical from
// src/CircleAI.Personal.Finance/PersonalFinancePrimitives.cs and the static
// domain-context constants from PersonalFinanceDomainContext.cs:
//   • FinanceAccount, FinanceTransaction, BudgetLine, MonthSummary — records
//   • IPersonalFinanceBoard   — accounts / transactions / budgets / summary
//   • InMemoryPersonalFinanceBoard — deterministic in-memory impl
//   • PersonalFinanceDomainContext — system-prompt snippet + flags
//
// The Companion-facing wrapper (PersonalFinanceCompanionAdapter) is intentionally
// NOT ported (see Healthcare.swift for the rationale).
//
// Porting notes:
//   • The C# record `Account` is renamed `FinanceAccount` so it does not collide
//     with `BankAccount` (Banking) in the flat Swift module.
//   • `decimal` → `Decimal`; `DateTimeOffset` → `Date`; `string? Note` → `String?`.
//   • `Record` on an unknown account throws → `PersonalFinanceError.unknownAccount`,
//     and (as in C#) adjusts the account balance by the transaction amount.
//   • Budgets are keyed case-insensitively (C# `StringComparer.OrdinalIgnoreCase`);
//     `Budgets` returns them ordered ascending by category.
//   • `ListForMonth` filters by account + calendar year/month of `atUtc`
//     (extracted with a UTC calendar for locale independence).
//   • `Summarise` buckets by category, `totalIn` = Σ positive amounts,
//     `totalOut` = −Σ negative amounts.

import Foundation

// MARK: - Records

/// A personal-finance account. (C# `Account` in CircleAI.Personal.Finance.)
public struct FinanceAccount: Sendable, Equatable, Codable {
    /// Stable identifier for the account.
    public let accountId: String
    /// Account name.
    public let name: String
    /// Current balance.
    public let balance: Decimal
    /// ISO currency code.
    public let currency: String

    public init(accountId: String, name: String, balance: Decimal, currency: String) {
        self.accountId = accountId
        self.name = name
        self.balance = balance
        self.currency = currency
    }
}

/// A personal-finance transaction (signed amount).
public struct FinanceTransaction: Sendable, Equatable, Codable {
    /// Stable identifier for the transaction.
    public let txId: String
    /// Identifier of the account the transaction posts to.
    public let accountId: String
    /// Signed amount (positive = money in, negative = money out).
    public let amount: Decimal
    /// Category (e.g. "groceries").
    public let category: String
    /// Optional note.
    public let note: String?
    /// UTC timestamp.
    public let atUtc: Date

    public init(txId: String, accountId: String, amount: Decimal, category: String,
                note: String?, atUtc: Date) {
        self.txId = txId
        self.accountId = accountId
        self.amount = amount
        self.category = category
        self.note = note
        self.atUtc = atUtc
    }
}

/// A monthly budget limit for a category.
public struct BudgetLine: Sendable, Equatable, Codable {
    /// Category the budget applies to.
    public let category: String
    /// Monthly spending limit.
    public let monthlyLimit: Decimal

    public init(category: String, monthlyLimit: Decimal) {
        self.category = category
        self.monthlyLimit = monthlyLimit
    }
}

/// A one-month roll-up of activity for an account.
public struct MonthSummary: Sendable, Equatable, Codable {
    /// Four-digit year.
    public let year: Int
    /// Month 1–12.
    public let month: Int
    /// Total money in (sum of positive amounts).
    public let totalIn: Decimal
    /// Total money out (absolute value of the sum of negative amounts).
    public let totalOut: Decimal
    /// Net amount per category.
    public let byCategory: [String: Decimal]

    public init(year: Int, month: Int, totalIn: Decimal, totalOut: Decimal, byCategory: [String: Decimal]) {
        self.year = year
        self.month = month
        self.totalIn = totalIn
        self.totalOut = totalOut
        self.byCategory = byCategory
    }
}

// MARK: - Errors

/// Errors thrown by the personal-finance board.
public enum PersonalFinanceError: Error, Equatable, CustomStringConvertible {
    /// `record` targeted an account id that is not known.
    case unknownAccount(String)

    public var description: String {
        switch self {
        case .unknownAccount(let id): return "Unknown account \(id)"
        }
    }
}

// MARK: - IPersonalFinanceBoard

/// Accounts, transactions, budgets, and monthly summaries for the
/// personal-finance vertical. A synchronous contract — implementations are
/// expected to be thread-safe.
public protocol IPersonalFinanceBoard: AnyObject, Sendable {
    /// Inserts (or replaces, by `accountId`) an account.
    func upsert(_ a: FinanceAccount)
    /// Returns the account with `id`, or `nil`.
    func getAccount(_ id: String) -> FinanceAccount?
    /// Records a transaction, adjusting the account balance. Throws when unknown.
    func record(_ t: FinanceTransaction) throws
    /// Transactions for `accountId` in the given calendar month.
    func listForMonth(accountId: String, year: Int, month: Int) -> [FinanceTransaction]
    /// Sets (or replaces, case-insensitively by category) a budget line.
    func setBudget(_ b: BudgetLine)
    /// Budget lines ordered ascending by category.
    var budgets: [BudgetLine] { get }
    /// Summarises `accountId` for the given calendar month.
    func summarise(accountId: String, year: Int, month: Int) -> MonthSummary
}

// MARK: - InMemoryPersonalFinanceBoard

/// Deterministic in-memory `IPersonalFinanceBoard`. All state is guarded by a
/// single `NSLock`; every accessor returns an immutable snapshot.
public final class InMemoryPersonalFinanceBoard: IPersonalFinanceBoard, @unchecked Sendable {
    private let lock = NSLock()
    private var accounts: [String: FinanceAccount] = [:]
    // Budget keys are lower-cased for case-insensitive replacement; the original
    // category casing is preserved inside the stored BudgetLine.
    private var budgetsByKey: [String: BudgetLine] = [:]
    private var txns: [FinanceTransaction] = []

    private static let utcCalendar: Calendar = {
        var c = Calendar(identifier: .gregorian)
        c.timeZone = TimeZone(identifier: "UTC")!
        c.locale = Locale(identifier: "en_US_POSIX")
        return c
    }()

    public init() {}

    public func upsert(_ a: FinanceAccount) {
        lock.lock(); defer { lock.unlock() }
        accounts[a.accountId] = a
    }

    public func getAccount(_ id: String) -> FinanceAccount? {
        lock.lock(); defer { lock.unlock() }
        return accounts[id]
    }

    public func record(_ t: FinanceTransaction) throws {
        lock.lock(); defer { lock.unlock() }
        guard let a = accounts[t.accountId] else { throw PersonalFinanceError.unknownAccount(t.accountId) }
        txns.append(t)
        accounts[t.accountId] = FinanceAccount(accountId: a.accountId, name: a.name,
                                               balance: a.balance + t.amount, currency: a.currency)
    }

    public func listForMonth(accountId: String, year: Int, month: Int) -> [FinanceTransaction] {
        lock.lock(); defer { lock.unlock() }
        return txns.filter {
            guard $0.accountId == accountId else { return false }
            let c = Self.utcCalendar.dateComponents([.year, .month], from: $0.atUtc)
            return c.year == year && c.month == month
        }
    }

    public func setBudget(_ b: BudgetLine) {
        lock.lock(); defer { lock.unlock() }
        budgetsByKey[b.category.lowercased()] = b
    }

    public var budgets: [BudgetLine] {
        lock.lock(); defer { lock.unlock() }
        return budgetsByKey.values.sorted { $0.category < $1.category }
    }

    public func summarise(accountId: String, year: Int, month: Int) -> MonthSummary {
        let rows = listForMonth(accountId: accountId, year: year, month: month)
        var byCat: [String: Decimal] = [:]
        for t in rows { byCat[t.category, default: 0] += t.amount }
        let inSum = rows.filter { $0.amount > 0 }.reduce(Decimal(0)) { $0 + $1.amount }
        let outSum = -rows.filter { $0.amount < 0 }.reduce(Decimal(0)) { $0 + $1.amount }
        return MonthSummary(year: year, month: month, totalIn: inSum, totalOut: outSum, byCategory: byCat)
    }
}

// MARK: - PersonalFinanceDomainContext

/// Static domain-context constants for the personal-finance vertical. Mirrors
/// `PersonalFinanceDomainContext` in PersonalFinanceDomainContext.cs.
public enum PersonalFinanceDomainContext {
    /// System-prompt snippet injected ahead of user turns in this domain.
    public static let systemPromptSnippet = "[DOMAIN: Personal.Finance] Personal finance coach. Help with monthly budgeting, emergency fund planning, debt snowball/avalanche strategy, savings goals, retirement planning basics, and investment options education. IMPORTANT: This is financial education, not advice. Recommend a registered financial planner for personalised investment advice. Compliance: FAIS Act, NCA, POPIA."
    /// Regulatory / compliance flags relevant to this domain.
    public static let complianceFlags: [String] = ["FAIS_Act_37_2002", "NCA", "POPIA", "Not_Financial_Advice"]
    /// Tools the Companion may suggest in this domain.
    public static let suggestedTools: [String] = ["budget_tracker", "spreadsheet", "calculator", "web_search"]
}

File: kotlin/src/main/kotlin/com/bhengubv/circleai/personalfinance/PersonalFinance.kt
// PersonalFinance.kt
//
// Kotlin port of CircleAI.Personal.Finance (PersonalFinancePrimitives.cs +
// PersonalFinanceDomainContext.cs + PersonalFinanceCompanionAdapter.cs) — the
// C# reference is the EXACT spec. Deterministic in-memory personal-finance
// store: accounts, transactions, budgets, and a monthly summary.
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`.
//   * C# `decimal` -> `java.math.BigDecimal`.
//   * C# `DateTimeOffset` -> `java.time.Instant`.
//   * C# `IReadOnlyDictionary<string,decimal>` -> `Map<String, BigDecimal>`.
//   * Accounts keyed Ordinal; budgets keyed OrdinalIgnoreCase (case-insensitive
//     category) -> a `ConcurrentHashMap` with lower-cased keys preserving the
//     original category on the value.
//   * `Record` throws on an unknown account and mutates the account balance;
//     transactions live in a plain list behind a lock.
//   * `ListForMonth` filters by account + year + month.
//   * `Budgets` orders by category ASC.
//   * `Summarise` groups signed amounts by category; TotalIn = Σ positive,
//     TotalOut = −Σ negative.

package com.bhengubv.circleai.personalfinance

import com.bhengubv.circleai.companion.CompanionContext
import com.bhengubv.circleai.companion.CompanionProactiveEvent
import com.bhengubv.circleai.companion.CompanionTurn
import com.bhengubv.circleai.companion.ICompanionSession
import com.bhengubv.circleai.companion.InterfaceKind
import kotlinx.coroutines.flow.Flow
import java.math.BigDecimal
import java.time.Instant
import java.time.ZoneOffset
import java.util.concurrent.ConcurrentHashMap

// =====================================================================
// Primitives (PersonalFinancePrimitives.cs)
// =====================================================================

/** A personal-finance account. Mirrors C# `Account`. */
data class Account(val accountId: String, val name: String, val balance: BigDecimal, val currency: String)

/** A recorded transaction (signed amount). Mirrors C# `FinanceTransaction`. */
data class FinanceTransaction(
    val txId: String,
    val accountId: String,
    val amount: BigDecimal,
    val category: String,
    val note: String?,
    val atUtc: Instant,
)

/** A per-category monthly budget cap. Mirrors C# `BudgetLine`. */
data class BudgetLine(val category: String, val monthlyLimit: BigDecimal)

/** A month's rolled-up summary. Mirrors C# `MonthSummary`. */
data class MonthSummary(
    val year: Int,
    val month: Int,
    val totalIn: BigDecimal,
    val totalOut: BigDecimal,
    val byCategory: Map<String, BigDecimal>,
)

/** Deterministic personal-finance board. Mirrors C# `IPersonalFinanceBoard`. */
interface IPersonalFinanceBoard {
    fun upsert(a: Account)
    fun getAccount(id: String): Account?
    fun record(t: FinanceTransaction)
    fun listForMonth(accountId: String, year: Int, month: Int): List<FinanceTransaction>
    fun setBudget(b: BudgetLine)
    val budgets: List<BudgetLine>
    fun summarise(accountId: String, year: Int, month: Int): MonthSummary
}

/** In-memory [IPersonalFinanceBoard]. Mirrors C# `InMemoryPersonalFinanceBoard`. */
class InMemoryPersonalFinanceBoard : IPersonalFinanceBoard {
    private val accounts = ConcurrentHashMap<String, Account>()
    // OrdinalIgnoreCase budget keying: lower-cased key, original category on value.
    private val budgets_ = ConcurrentHashMap<String, BudgetLine>()
    private val txns = mutableListOf<FinanceTransaction>()
    private val lock = Any()

    override fun upsert(a: Account) { accounts[a.accountId] = a }
    override fun getAccount(id: String): Account? = accounts[id]

    override fun record(t: FinanceTransaction) {
        if (!accounts.containsKey(t.accountId)) throw IllegalStateException("Unknown account ${t.accountId}")
        synchronized(lock) {
            txns.add(t)
            val a = accounts.getValue(t.accountId)
            accounts[t.accountId] = a.copy(balance = a.balance + t.amount)
        }
    }

    override fun listForMonth(accountId: String, year: Int, month: Int): List<FinanceTransaction> =
        synchronized(lock) {
            txns.filter {
                it.accountId == accountId &&
                    it.atUtc.atZone(ZoneOffset.UTC).year == year &&
                    it.atUtc.atZone(ZoneOffset.UTC).monthValue == month
            }
        }

    override fun setBudget(b: BudgetLine) { budgets_[b.category.lowercase()] = b }

    override val budgets: List<BudgetLine>
        get() = budgets_.values.sortedBy { it.category }

    override fun summarise(accountId: String, year: Int, month: Int): MonthSummary {
        val rows = listForMonth(accountId, year, month)
        val byCat = rows.groupBy { it.category }
            .mapValues { (_, v) -> v.fold(BigDecimal.ZERO) { acc, t -> acc + t.amount } }
        val inSum = rows.filter { it.amount > BigDecimal.ZERO }
            .fold(BigDecimal.ZERO) { acc, t -> acc + t.amount }
        val outSum = rows.filter { it.amount < BigDecimal.ZERO }
            .fold(BigDecimal.ZERO) { acc, t -> acc + t.amount }.negate()
        return MonthSummary(year, month, inSum, outSum, byCat)
    }
}

// =====================================================================
// DomainContext (PersonalFinanceDomainContext.cs)
// =====================================================================

/** Static domain context for Personal.Finance. Mirrors C# `PersonalFinanceDomainContext`. */
object PersonalFinanceDomainContext {
    const val SYSTEM_PROMPT_SNIPPET: String =
        "[DOMAIN: Personal.Finance] Personal finance coach. Help with monthly budgeting, emergency fund " +
            "planning, debt snowball/avalanche strategy, savings goals, retirement planning basics, and " +
            "investment options education. IMPORTANT: This is financial education, not advice. Recommend a " +
            "registered financial planner for personalised investment advice. Compliance: FAIS Act, NCA, POPIA."

    val complianceFlags: List<String> = listOf("FAIS_Act_37_2002", "NCA", "POPIA", "Not_Financial_Advice")

    val suggestedTools: List<String> = listOf("budget_tracker", "spreadsheet", "calculator", "web_search")
}

// =====================================================================
// CompanionAdapter (PersonalFinanceCompanionAdapter.cs)
// =====================================================================

/**
 * Wraps an [ICompanionSession] with the Personal.Finance snippet + helpers.
 * Mirrors C# `PersonalFinanceCompanionAdapter`.
 */
class PersonalFinanceCompanionAdapter(private val inner: ICompanionSession) : ICompanionSession {
    override val sessionId: String get() = inner.sessionId
    override val identityId: String get() = inner.identityId
    override val interfaceKind: InterfaceKind get() = inner.interfaceKind
    override val history: List<CompanionTurn> get() = inner.history
    override val proactiveEvents: Flow<CompanionProactiveEvent> get() = inner.proactiveEvents

    override fun getContext(): CompanionContext = inner.getContext()
    override suspend fun refreshContextAsync() = inner.refreshContextAsync()
    override suspend fun signalFeedbackAsync(positive: Boolean, note: String?) =
        inner.signalFeedbackAsync(positive, note)
    override fun close() = inner.close()

    override suspend fun sendAsync(message: String): String = inner.sendAsync(enrich(message))
    override fun streamAsync(message: String): Flow<String> = inner.streamAsync(enrich(message))
    override suspend fun agentAsync(instruction: String): String = inner.agentAsync(enrich(instruction))

    private fun enrich(m: String): String = "${PersonalFinanceDomainContext.SYSTEM_PROMPT_SNIPPET}\n\n$m"

    suspend fun buildBudgetAsync(income: String, expenses: String): String =
        inner.agentAsync("Build a monthly budget. Income: $income. Expenses: $expenses. Apply the 50/30/20 rule, identify savings opportunities, and flag over-spending categories.")

    suspend fun createDebtPlanAsync(debts: String): String =
        inner.agentAsync("Create a debt elimination plan using the avalanche method (highest interest first):\n$debts\nShow monthly payment schedule, total interest saved, and debt-free date.")

    suspend fun analyseSpendingAsync(categoryBreakdown: String, monthlyIncome: String): String =
        inner.agentAsync("Analyse spending $categoryBreakdown against income $monthlyIncome. Identify 2 leaks + a realistic redirect target.")

    suspend fun designSavingsGoalAsync(goal: String, targetAmount: BigDecimal, monthsAvailable: Int): String =
        inner.agentAsync("Plan to save $targetAmount for '$goal' in $monthsAvailable months. Monthly target + behavioural commitment device.")

    suspend fun explainTaxImpactAsync(scenario: String, jurisdiction: String): String =
        inner.agentAsync("Explain tax impact of: $scenario in $jurisdiction. Likely treatment, paperwork, optimisation lever. Not tax advice.")

    suspend fun reviewInvestmentMixAsync(portfolio: String, riskAppetite: String, horizonYears: Int): String =
        inner.agentAsync("Review investment mix: $portfolio against $riskAppetite appetite, $horizonYears-year horizon. Coverage, concentration, fee drag.")
}
